modules/@apostrophecms/i18n/index.js

```diff
@@ -62,6 +62,59 @@ module.exports = {
         addModal() {
           self.addLocalizeModal();
         }
+      },
+      '@apostrophecms/page:beforeSend': {
+        // Developers can link to alternate locales by iterating over
+        // `data.localizations` in any page template. Each element always has
+        // `locale`, `label` and `homePageUrl` properties. Each element also has an
+        // `available` property; if true, the current context document is available
+        // in that locale, `title` and a small number of other document properties are
+        // populated, and `_url` redirects to the context document in that locale.
+        //
+        // The array is provided in the order in which locales are configured.
+        // The current locale is included and has the property `current: true`.
+        async addLocalizations(req) {
+          const context = req.data.piece || req.data.page;
+          if (!context) {
+            return;
+          }
+          const manager = self.apos.doc.getManager(context.type);
+          if (!manager.isLocalized()) {
+            return;
+          }
+          const localizations = await self.apos.doc.db.find({
+            aposDocId: context.aposDocId,
+            aposMode: req.mode
+          }).project({
+            type: 1,
+            title: 1,
+            slug: 1,
+            aposLocale: 1,
+            aposMode: 1,
+            visibility: 1,
+            docPermissions: 1
+          }).toArray();
+          req.data.localizations = [];
+          for (const name of Object.keys(self.locales)) {
+            const localeReq = self.apos.util.cloneReq(req, {
+              locale: name
+            });
+            self.setPrefixUrls(localeReq);
+            const doc = localizations.find(doc => doc.aposLocale.split(':')[0] === name);
+            if (doc && self.apos.permission.can(req, 'view', doc)) {
+              doc.available = true;
+              doc._url = `${req.prefix}${manager.action}/${context._id}/locale/${name}`;
```

Comment:
With 90 locales, like some of our enterprise clients, and the need to potentially drag querying for locale-specific piece pages into the calculation, this query would bring the site to its knees. Which is why we also defer the calculation until needed this way in A2. The links we're talking about will more typically be in a dropdown menu, possibly not even hoverable links in the usual way.

We could use slugs rather than ids in this API, although Apostrophe document _ids are not in any way meant to be used as secrets (they are often publicly exposed in headless APIs).